### PR TITLE
Add unit tests for `StreamPeer` and `StreamPeerBuffer`

### LINKED ISSUE #95783: `StreamPeerBuffer` fails to return only a part of the string previously stored on it
### Tested versions

- Reproducible in: 4.3.stable

### System information

Godot v4.3.stable - macOS 14.6.1 - GLES3 (Compatibility) - Intel(R) Iris(TM) Plus Graphics OpenGL Engine - Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz (8 Threads)

### Issue description

If you put a string or an utf8 string in a `StreamPeerBuffer` and then you try to get just a part of it you will get an empty or corrupted string.

### Steps to reproduce

Here is a small script to reproduce the issue.
```gdscript
extends Node


# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	var buff = StreamPeerBuffer.new()
	buff.put_string("Hello, World!!!")
	buff.seek(0)
	print("->" + buff.get_string(5) + "<-")
	
	buff.clear()
	buff.put_utf8_string("Hello✩, World✩!!!")
	buff.seek(0)
	print("->" + buff.get_utf8_string(8) + "<-")
```

This script will print:
```
-><-
-><-
```

### Minimal reproduction project (MRP)

[streampeerbuffergetpartialstringbug.zip](https://github.com/user-attachments/files/16654388/streampeerbuffergetpartialstringbug.zip)

### PR DESCRIPTION
**Pull request by pafuent**
This PR aims to help "fix" #43440

Adding unit tests for `StreamPeer` and `StreamPeerBuffer`

### CODE REVIEW

**Comment by pafuent** [comment]
Because the discussion around if the bug referenced by this PR is a bug or not, I can extract all the other unit test in another PR to get that one move faster and get merged. Thoughts?

**Comment by pafuent** [comment]
I decided to remove #95783 fix and keep this PR as only adding tests

**Comment by pafuent** [comment]
@timothyqiu Could you please remove `breaks compat` label?
The PR only add tests now.

**Comment by pafuent** [comment]
@Calinou Can I get a review now that the PR is only adding Unit Tests?

**Review by Geometror — COMMENTED** [review]
To be honest, I wasn't sure about the separation of the StreamPeer and StreemPeerBuffer tests, but after looking into this a bit longer I now think it makes sense. Maybe adding a StreamPeerDummy/Mock would be the most correct way to test them apart, but I'd leave it up to you if you want to do that.

Either way, needs to be squashed into one commit.

File: tests/core/io/test_stream_peer_buffer.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	CHECK_EQ(0, spb->get_position());
+	CHECK_EQ(42, spb->get_available_bytes());
+
+	spb->seek(21);
+	CHECK_EQ(42, spb->get_size());
+	CHECK_EQ(21, spb->get_position());
+	CHECK_EQ(21, spb->get_available_bytes());
+}
+
+TEST_CASE("[StreamPeerBuffer] Get underling data array") {
```
Comment (AThousandShips) [inline tests/core/io/test_stream_peer_buffer.h:96]:
```suggestion
TEST_CASE("[StreamPeerBuffer] Get underlying data array") {
```

File: tests/core/io/test_stream_peer_buffer.h [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	spb->put_u8(third);
+
+	Vector<uint8_t> data_array = spb->get_data_array();
+
+	CHECK_EQ(first, data_array[0]);
+	CHECK_EQ(second, data_array[1]);
+	CHECK_EQ(third, data_array[2]);
+}
+
+TEST_CASE("[StreamPeerBuffer] Set underling data array") {
```
Comment (AThousandShips) [inline tests/core/io/test_stream_peer_buffer.h:114]:
```suggestion
TEST_CASE("[StreamPeerBuffer] Set underlying data array") {
```

File: tests/core/io/test_stream_peer.h [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	spb.instantiate();
+
+	CHECK_EQ(false, spb->is_big_endian_enabled());
+}
+
+TEST_CASE("[StreamPeer] Get and sets through StreamPeerBuffer") {
+	Ref<StreamPeerBuffer> spb;
+	spb.instantiate();
+
+	SUBCASE("a int8_t value") {
```
Comment (AThousandShips) [inline tests/core/io/test_stream_peer.h:50]:
```suggestion
	SUBCASE("A int8_t value") {
```
For all

**Comment by pafuent** [comment]
@Geometror 
> To be honest, I wasn't sure about the separation of the StreamPeer and StreemPeerBuffer tests, but after looking into this a bit longer I now think it makes sense. Maybe adding a StreamPeerDummy/Mock would be the most correct way to test them apart, but I'd leave it up to you if you want to do that.

In order to add a mock I prefer to wait until a definition on #95224 where I added a Mocking library. If I get that PR merged with the mocking library I will be glad to revisit this tests and use a mock.

**Review by Calinou — APPROVED** [review]

File: tests/core/io/test_stream_peer.h [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+#include "core/io/stream_peer.h"
+#include "tests/test_macros.h"
+
+namespace TestStreamPeer {
+
+TEST_CASE("[StreamPeer] Initialization through StreamPeerBuffer") {
+	Ref<StreamPeerBuffer> spb;
+	spb.instantiate();
+
+	CHECK_EQ(false, spb->is_big_endian_enabled());
```
Comment (AThousandShips) [inline tests/core/io/test_stream_peer.h:43]:
```suggestion
	CHECK_EQ(spb->is_big_endian_enabled(), false);
```
For all, value to check against on the right

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**